src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1728,17 +1728,19 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
                 keyword = ConvertToKeyword(this.EatToken());
             }
 
+            bool isExtension = keyword.Kind == SyntaxKind.ExtensionKeyword;
             var outerSaveTerm = _termState;
-            _termState |= TerminatorState.IsEndOfRecordOrClassOrStructOrInterfaceSignature;
+            _termState |= TerminatorState.IsEndOfRecordOrClassOrStructOrInterfaceOrExtensionSignature;
 
             var saveTerm = _termState;
             _termState |= TerminatorState.IsPossibleAggregateClauseStartOrStop;
 
-            var name = this.ParseIdentifierToken();
+            SyntaxToken? name = (isExtension && this.CurrentToken.Kind != SyntaxKind.IdentifierToken) ? null : this.ParseIdentifierToken();
```

Comment:
nice.  consider stating that we will parse an identifier if present for extensions, but will error later in binding/symbols.